### PR TITLE
Replace boost::random with std::random

### PR DESCRIPTION
**Pull request by martyngigg**
**Description of work**

While upgrading boost as part of #22088 it was observed that the output of `boost::random::normal_distribution` has been slightly tweaked again producing some test failures. This work migrates everything to use classes from `std` `<random>` in the hope that we can standardize on this implementation and have fewer tests which need to check versions of boost etc.

As part of this work it was noticed that `std::normal_distribution` produces matching output on MSVC and clang but not [gcc](https://stackoverflow.com/questions/38532927/why-gcc-and-msvc-stdnormal-distribution-are-different). I have decided that to decrease the overhead in maintaining tests that the `normal_distribution` implementation that is part of clang would be brought into `Kernel` and used in place of the `std` version.

Many tests relied on sample from a uniform or normal distribution and this done slightly differently in the `std` versions but our tests don't perform enough iterations for the differences to be unobservable. The tests have been updated accordingly.

Note that I have also snuck in a small fix for some python tests that started failing after a recent update on Ubuntu - please let me know if you would like those separated out.

**To test:**

* Code review and tests should pass.

*No issue*

**Release Notes** 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Kernel/inc/MantidKernel/uniform_int_distribution.h [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+ */
+#include <limits>
+#include <random>
+#include <type_traits>
+
+// Implementation of uniform_int_distribution taken from llvm/libcxx at
+// https://github.com/llvm-mirror/libcxx/blob/master/include/algorithm#L2965
+//
+// Each of our supported platforms has the header but minor differences in the
+// results yeild problems consistency of our tests across platorms.
```
Comment (SimonHeybrock) [inline Framework/Kernel/inc/MantidKernel/uniform_int_distribution.h]: yield

**Review by SimonHeybrock — CHANGES_REQUESTED** [review]
Code looks good and I think this is a sensible way to solve the described consistency issues 👍.
Changes also include some cleanup and re-enabled tests. 👍 

Only a couple of minor questions/comments:

File: Framework/Kernel/inc/MantidKernel/normal_distribution.h [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+  File change history is stored at: <https://github.com/mantidproject/mantid>
+  Code Documentation is available at: <http://doxygen.mantidproject.org>
+ */
+#include "MantidKernel/System.h"
+#include <ios>
```
[the remark this answers is not in the document]
Comment (martyngigg) [reply Framework/Kernel/inc/MantidKernel/normal_distribution.h]: Good question. As I was pulling the the header from `llvm` I wanted to keep it as similar as possible. Although after looking over it again we're going to get it included anyway due to the inclusion of `<random>` so I should probably remove here.

[the remark this answers is not in the document]
Comment (martyngigg) [reply Framework/Kernel/inc/MantidKernel/normal_distribution.h]: I think we currently only use it from a few places in the code base.

File: Framework/CurveFitting/src/FuncMinimizers/FABADAMinimizer.cpp [hunk 3]
```diff
@@ -355,13 +352,9 @@ void FABADAMinimizer::finalize() {
 * @return :: the step
 */
 double FABADAMinimizer::gaussianStep(const double &jump) {
```
[the remark this answers is not in the document]
Comment (martyngigg) [reply Framework/CurveFitting/src/FuncMinimizers/FABADAMinimizer.cpp]: I'm afraid I don't. The algorithm is quite a complex method for Bayesian fitting as far as I understand it. @mantid-roman Could you offer any insight here?

[the remark this answers is not in the document]
Comment (mantid-roman) [reply Framework/CurveFitting/src/FuncMinimizers/FABADAMinimizer.cpp]: I think that the use of `mt` is wrong here. The effective random number generator here is the seed expression.

[the remark this answers is not in the document]
Comment (martyngigg) [reply Framework/CurveFitting/src/FuncMinimizers/FABADAMinimizer.cpp]: @mantid-roman Due you mean essentially that the return value should just be the seed expression?

[the remark this answers is not in the document]
Comment (mantid-roman) [reply Framework/CurveFitting/src/FuncMinimizers/FABADAMinimizer.cpp]: No. `mt` should be instantiated outside the function and passed to it as a parameter (or a global variable). `mt` generates uniform random numbers using some sophisticated algorithm and those are used to pick values from the normal distribution. But what happens currently is that this sophisticated algorithm is effectively replaced by the expression `123 * (int(m_counter) + 45 * int(jump))  + 14 * int(time_t())`. Sorry I wasn't clear in my first comment.

[the remark this answers is not in the document]
Comment (martyngigg) [reply Framework/CurveFitting/src/FuncMinimizers/FABADAMinimizer.cpp]: No problem. Thanks for clarifying.

File: Framework/Crystal/test/NormaliseVanadiumTest.h [hunk 4]
```diff
@@ -172,14 +137,9 @@ class NormaliseVanadiumTest : public CxxTest::TestSuite {
     TS_ASSERT(ws);
     if (!ws)
       return;
-    TS_ASSERT_DELTA(ws->y(5050)[5], 7.7142, 0.01);
+    TS_ASSERT_DELTA(ws->y(5050)[5], 17, 0.0001);
```
[the remark this answers is not in the document]
Comment (martyngigg) [reply Framework/Crystal/test/NormaliseVanadiumTest.h]: It seems to be entirely down to the way the random numbers are sampled from the flat distribution. If I make a one line change and use `boost::random::uniform_real_distribution` and leave the other changes alone then I recover the original number. I assume using a much higher number of events would eventually yield a converging result for both implementations.

File: Framework/CurveFitting/src/FuncMinimizers/FABADAMinimizer.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (martyngigg) [reply Framework/CurveFitting/src/FuncMinimizers/FABADAMinimizer.cpp]: I think I would prefer to have a separate issue for replacing this. It's going to require changes to allow a seed to be passed in and then we'll need to make sure we test it well. I've created #22317 to look at this.

**Review by SimonHeybrock — APPROVED** [review]

**Comment by rosswhitfield** [comment]
Can you check the VesuvioCorrectionsTest systemtest? I'm guessing it's related. #19782

**Comment by martyngigg** [comment]
@rosswhitfield Will do. I suspect it is the same problem too. It'll be nice for this not to be an issue in the future!

**Comment by martyngigg** [comment]
@rosswhitfield I've checked and updated the Vesuvio tests and they now seem to pass for on RHEL7, Ubuntu and Windows. Hopefully this means we have one less source of discrepancy between the OSs.

**Review by SimonHeybrock — APPROVED** [review]

**Comment by martyngigg** [comment]
@mantidproject/gatekeepers Could we get this merged pretty soon? It updates a fair few files across the framework directory.

**Comment by rosswhitfield** [comment]
Thanks @martyngigg

**Pull request merged**